### PR TITLE
Build Report view

### PR DESCRIPTION
**Pull request by mtrive**
At the moment, the only tool that allows you to inspect Build Reports is [Build Report Inspector](https://github.com/Unity-Technologies/BuildReportInspector) which is not currently being developed.

With this PR, Project Auditor features a Build Report view to list all files included in the build.

build-report

This is similar to the Build Report Inspector "SourceAssets" view.

build-report-inspector

The main difference is that the data is sortable, searchable and exportable.

Future improvements:
- Group by type and show aggregate size. Alternatively, there could be a TreeView Map similar to the one in the memory profiler.
- Show assets dependencies panel
- Add build steps panel

### CODE REVIEW

**Comment by berniegp** [comment]
I'll try to have a look at this Monday. Couple of things ahead of that:
 - Would it be possible to have a "human readable size" for the size column instead of bytes like in the Build Report Inspector.
 - This only shows assets that are part of the build itself right? In other words, not AssetBundles? It's fine either way; I just want to understand the intended scope.

**Comment by mtrive** [comment]
> * Would it be possible to have a "human readable size" for the size column instead of bytes like in the Build Report Inspector.
It's certainly possible. It's a special case compared to everything else but it make sense to support.

> * This only shows assets that are part of the build itself right? In other words, not AssetBundles? It's fine either way; I just want to understand the intended scope.
that's correct but something similar could be built for AssetBundles. Maybe by re-using some of [this](https://github.com/sbeck3d/com.unity.asset-bundle-analyzer).

File: Editor/Auditors/BuildAuditor.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            if (!Directory.Exists(s_BuildReportDir))
+                Directory.CreateDirectory(s_BuildReportDir);
+
+            var date = File.GetLastWriteTime(s_LastBuildReportPath);
+            var assetPath = s_BuildReportDir + "/Build_" + date.ToString("yyyy-MM-dd-HH-mm-ss") + ".buildreport";
+
+            if (!File.Exists(assetPath))
+            {
+                File.Copy("Library/LastBuild.buildreport", assetPath, true);
+                AssetDatabase.ImportAsset(assetPath);
```
Comment (berniegp) [inline Editor/Auditors/BuildAuditor.cs:127]: Should this stay in the project as an asset or get cleaned up after analysis?

File: Editor/Auditors/BuildAuditor.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        public void Audit(Action<ProjectIssue> onIssueFound, Action onComplete, IProgressBar progressBar = null)
+        {
+#if UNITY_2019_4_OR_NEWER
+            var buildReport = GetBuildReport();
+            if (buildReport != null)
+            {
+                foreach (var packedAsset in buildReport.packedAssets)
+                {
+                    var dict = new Dictionary<GUID, List<PackedAssetInfo>>();
```
Comment (berniegp) [inline Editor/Auditors/BuildAuditor.cs:65]: Could you add a comment to explain why there can be more than one content generated per source asset GUID and why we want to aggregate them together? It's not obvious to me with my limited knowledge of the build process.

**Review by berniegp — CHANGES_REQUESTED** [review]
I didn't fully understand the refactor in `ProjectAuditorWindow` (the diff is a bit hard to follow), but it looks ok at first glance.

General comment note related to this PR: how far back does Project Auditor support Unity versions? Might want to do a pass and remove code paths that support versions out of LTS (i.e. 2018) to simplify the code at some point.

File: Editor/UI/BuildReportView.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        private static string FormatSize(ulong size)
+        {
+            if (size < 1024)
+                return size + " B";
+            if (size < 1024 * 1024)
+                return (size / 1024.00).ToString("F2") + " KB";
+            if (size < 1024 * 1024 * 1024)
+                return (size / (1024.0 * 1024.0)).ToString("F2") + " MB";
+            return (size / (1024.0 * 1024.0 * 1024.0)).ToString("F2") + " GB";
+        }
```
[the remark this answers is not in the document]
Comment (mtrive) [reply Editor/UI/BuildReportView.cs:40]: Good point. I actually did on a separate branch. I will need to use _FormatSize_ to make the size in the UI more user-friendly (as you suggested)

File: Editor/UI/BuildReportView.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            EditorGUILayout.LabelField("Started at: ", report.summary.buildStartedAt.ToString());
+            EditorGUILayout.LabelField("Ended at: ", report.summary.buildEndedAt.ToString());
+            EditorGUILayout.LabelField("Total Time: ", FormatTime(report.summary.totalTime));
+#if UNITY_2019_3_OR_NEWER
+            EditorGUILayout.LabelField("Total Size: ", FormatSize(report.summary.totalSize));
+            EditorGUILayout.LabelField("Build Result: ", report.summary.result.ToString());
+#else
+            EditorGUILayout.LabelField("Total Size: ", FormatSize(report.summary.totalSize));
+            EditorGUILayout.LabelField("Build Result: ", report.summary.result.ToString());
+#endif
```
[the remark this answers is not in the document]
Comment (mtrive) [reply Editor/UI/BuildReportView.cs:23]: oh dear, I don't know. That was a copy&paste from Build Report Inspector. Will fix.

File: Editor/Auditors/BuildAuditor.cs (hunk 1 above)
Comment (mtrive) [reply Editor/Auditors/BuildAuditor.cs:127]: Good question. Build Report Inspector does that so you can _inspect_ previously saved reports. Since this is not possible in Project Auditor, perhaps you are right, we should clean things up.

File: Editor/Auditors/BuildAuditor.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        static readonly IssueLayout k_IssueLayout = new IssueLayout
+        {
+            category = IssueCategory.BuildFiles,
+            properties = new[]
+            {
+                new PropertyDefinition { type = PropertyType.Description, name = "Source Asset"},
+                new PropertyDefinition { type = PropertyType.FileType, name = "Type"},
+                new PropertyDefinition { type = PropertyType.Custom, format = PropertyFormat.Integer, name = "Size (bytes)", longName = "Size (bytes) in the Build"},
+                new PropertyDefinition { type = PropertyType.Path, name = "Path"},
+                new PropertyDefinition { type = PropertyType.Custom + 1, format = PropertyFormat.String, name = "Build File"}
```
[the remark this answers is not in the document]
Comment (mtrive) [reply Editor/Auditors/BuildAuditor.cs:30]: I am using the _com.unity.coding_ package for code formatting so my assumption is that this is consistent with our standards.

File: Editor/Auditors/BuildAuditor.cs (hunk 2 above)
Comment (mtrive) [reply Editor/Auditors/BuildAuditor.cs:66]: done: https://github.com/Unity-Technologies/ProjectAuditor/pull/24/commits/390f54b9841757a05cda38551db5948bdff1fcd5

File: Editor/UI/BuildReportView.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtrive) [reply Editor/UI/BuildReportView.cs]: fixed

File: Editor/UI/BuildReportView.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtrive) [reply Editor/UI/BuildReportView.cs]: Done. This should not appear anymore in the updated PR.

File: Editor/Auditors/BuildAuditor.cs (hunk 1 above)
Comment (mtrive) [reply Editor/Auditors/BuildAuditor.cs:128]: I looked at deleting the asset and there is a problem: If we delete it, the _BuildReport_ object will be released from memory. Since the BuildReport summary info is accessed every frame, we would end up re-creating and deleting the asset every frame. I think I would leave it as is for the time being. I will revisit it as part of the load/save feature.

**Comment by mtrive** [comment]
> I didn't fully understand the refactor in `ProjectAuditorWindow` (the diff is a bit hard to follow), but it looks ok at first glance.
> 
> General comment note related to this PR: how far back does Project Auditor support Unity versions? Might want to do a pass and remove code paths that support versions out of LTS (i.e. 2018) to simplify the code at some point.

Yes, it would be nice to simplify all code paths.

In theory it should be compatible all the way back to 2017 or oven 5.6, however, we only have automated test coverage for 2018 or greater. Testing in previous versions is a manual process so, things might break every now and then.

**Comment by mtrive** [comment]
@berniegp follow-up on displaying the size as KB/MB: I will address that separately as I consider that a shared utility.

File: Editor/Auditors/BuildAuditor.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (berniegp) [reply Editor/Auditors/BuildAuditor.cs:30]: I got this reply on Slack:

> I think you’re hitting a known issue. IIRC, it should work if you add the redundant () for the c-tor of those types.

Could you give it a try and I can reply to him? Or you can reply as well if you like. It's in the thread I linked in the PA channel.

File: Editor/Auditors/BuildAuditor.cs (hunk 2 above)
Comment (berniegp) [reply Editor/Auditors/BuildAuditor.cs:66]: Thanks!

**Review by berniegp — APPROVED** [review]
Looks good to me! There's only the super important curly brace issue pending then :)
